compiler/rustc_expand/src/proc_macro_server.rs

```diff
@@ -49,158 +49,179 @@ impl ToInternal<token::Delimiter> for Delimiter {
     }
 }
 
-impl FromInternal<(TreeAndSpacing, &'_ mut Vec<Self>, &mut Rustc<'_, '_>)>
-    for TokenTree<Group, Punct, Ident, Literal>
+impl FromInternal<(TokenStream, &mut Rustc<'_, '_>)>
+    for Vec<TokenTree<TokenStream, Span, Ident, Literal>>
 {
-    fn from_internal(
-        ((tree, spacing), stack, rustc): (TreeAndSpacing, &mut Vec<Self>, &mut Rustc<'_, '_>),
-    ) -> Self {
+    fn from_internal((stream, rustc): (TokenStream, &mut Rustc<'_, '_>)) -> Self {
         use rustc_ast::token::*;
 
-        let joint = spacing == Joint;
-        let Token { kind, span } = match tree {
-            tokenstream::TokenTree::Delimited(span, delim, tts) => {
-                let delimiter = pm::Delimiter::from_internal(delim);
-                return TokenTree::Group(Group { delimiter, stream: tts, span, flatten: false });
-            }
-            tokenstream::TokenTree::Token(token) => token,
-        };
-
-        macro_rules! tt {
-            ($ty:ident { $($field:ident $(: $value:expr)*),+ $(,)? }) => (
-                TokenTree::$ty(self::$ty {
-                    $($field $(: $value)*,)+
-                    span,
-                })
-            );
-            ($ty:ident::$method:ident($($value:expr),*)) => (
-                TokenTree::$ty(self::$ty::$method($($value,)* span))
-            );
-        }
-        macro_rules! op {
-            ($a:expr) => {
-                tt!(Punct::new($a, joint))
+        let mut cursor = stream.into_trees();
+        let mut trees = Vec::new();
```

Comment:
Might be interesting to see if `stream` can be asked for a could of its child `tokenstream::TokenTree`s, which won't necessarily be accurate to the `proc_macro` `TokenTree`s, but if rounded up to a power of 2, it could remove all but one reallocation, on average.

Maybe leave for future PR?

Reply:
It's pretty easy to get a count of child `TokenTree`s, there's just a `.len()` getter on `TokenStream`: https://github.com/rust-lang/rust/blob/2f3ddd9f594adf9773547aa7cedb43c4ac8ffd2f/compiler/rustc_ast/src/tokenstream.rs#L398-L400

I'll add something simple like that because it's pretty easy to do in this patch.